### PR TITLE
Feature: specify a modifier's pdf type

### PR DESCRIPTION
**Pull request by kratsg**
# Description

This PR attempts to optimize the internal calculations being made by grouping the modifiers into *meta-modifiers* instead to perform the calculations. This is discussed somewhat heavily in #150.

# Checklist Before Requesting Approver

- [ ] Tests are passing
- [ ] "WIP" removed from the title of the pull request

### CODE REVIEW

File: pyhf/modifiers/__init__.py [hunk 1]
```diff
@@ -93,27 +116,28 @@ def modifier(*args, **kwargs):
     name = kwargs.pop('name', None)
     constrained = bool(kwargs.pop('constrained', False))
     shared = bool(kwargs.pop('shared', False))
+    pdf_type = str(kwargs.pop('pdf_type', 'normal'))
```
Comment (lukasheinrich) [inline pyhf/modifiers/__init__.py]: we hardcode the default here as well as in `add_to_registry` , should we pick one or the other/

one way is to have 

`    pdf_type = str(kwargs.pop('pdf_type'))`
in the decorator and 

```
def ..(...pdf_type = None)
pdf_type = pdf_type or 'normal'
```

in the function

File: pyhf/modifiers/normsys.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ def expected_data(self, pars):
 
     def pdf(self, a, alpha):
         tensorlib, _ = get_backend()
-        return tensorlib.normal(a, alpha, tensorlib.astensor([1]))
+        return getattr(tensorlib, self.pdf_type)(a, alpha, tensorlib.astensor([1]))
```
Comment (lukasheinrich) [inline pyhf/modifiers/normsys.py]: here we see that the signature of the pdf_type obviously depends on the pdf used, we can just trust that this is correct and introduce checks later?

File: pyhf/modifiers/__init__.py (hunk 1 above)
Comment (kratsg) [reply pyhf/modifiers/__init__.py]: `pdf_type = str(kwargs.pop('pdf_type'))` fails if `pdf_type` isn't specified.

File: pyhf/modifiers/normsys.py (hunk 2 above)
Comment (kratsg) [reply pyhf/modifiers/normsys.py]: Good point, I can always grab the current tensorlib and validate at the time of modifier definition.

Comment (lukasheinrich) [reply pyhf/modifiers/normsys.py]: how woud you validate it? seems like you'd need to analyze the implementation of `def pdf` to check if the signature called matches the signature that's defined

File: pyhf/modifiers/__init__.py (hunk 1 above)
Comment (lukasheinrich) [reply pyhf/modifiers/__init__.py]: just realized we're doing the same with the kwargs above -- it's not a big issue so feel free to ignore.

**Comment by lukasheinrich** [comment]
the current diff LGTM, should we keep this open until we see an actual improvement in speed through the use of meta-modifiers or do you want to merge it in @kratsg

**Comment by kratsg** [comment]
> should we keep this open until we see an actual improvement in speed through the use of meta-modifiers or do you want to merge it in @kratsg

multiple small diffs is better. I'll rename this PR.

**Comment by kratsg** [comment]
> @kratsg still [WIP]?

My bad.

**Pull request merged**